03_letters/saurabhvyas9/main_test.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMainOutput(t *testing.T) {
+	// Given
+	var buf bytes.Buffer
+	out = &buf
+
+	// When
+	main()
+
+	// Then
+	expected := strconv.Quote("a:2\nb:1\n")
+	actual := strconv.Quote(buf.String())
+	assert.Equalf(t, expected, actual, "Letter frequency failed.")
+}
+
+var testCases = map[string]struct {
```

Comment:
It is nice to see a lot of test cases, but be wary of adding too many. You need to consider whether some test cases are adding any additional testing. For example, the "foreign language" and "emoticons" one are really testing the same thing - it is just a sequence of non-ascii runes. Likewise, your code makes no distinction between alpha, numeric, spaces, etc, so adding tests for those different cases is not testing your code any more widely.

This matters when it comes to maintenance. As code grows, if something in the function changes, possibly all the test cases need to be updated. This could add unnecessary work which becomes tedious after a while.

Next time, try to consider the minimum number of test cases you can come up with that exercise the various branches and boundaries in your code that completely covers it, but no more.

Reply:
Makes sense.